File: daily/74-Search-a-2D-Matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    ROWS, COLS := len(matrix), len(matrix[0])

    top, bottom := 0, ROWS - 1
    for top <= bottom {
        mid := (top + bottom) / 2
        if matrix[mid][COLS - 1] < target {
            top = mid + 1
        } else if matrix[mid][0] > target {
            bottom = mid - 1
        } else {
            break
        }
    }

    if top > bottom {
        return false
    }

    row := (top + bottom) / 2
    left, right := 0, COLS - 1
    for left <= right {
        mid := (left + right) / 2
        if matrix[row][mid] == target {
            return true
        } else if matrix[row][mid] > target {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    return false
}

File: daily/74-Search-a-2D-Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROWS, COLS = len(matrix), len(matrix[0])

        top, bottom = 0, ROWS - 1
        while top <= bottom:
            mid = (top + bottom) // 2
            if matrix[mid][-1] < target:
                top = mid + 1
            elif matrix[mid][0] > target:
                bottom = mid - 1
            else:
                break
        if not (top <= bottom):
            return False

        row = (top + bottom) // 2
        left, right = 0, COLS - 1
        while left <= right:
            mid = (left + right) // 2
            if matrix[row][mid] == target:
                return True
            elif matrix[row][mid] > target:
                right = mid - 1
            else:
                left = mid + 1

        return False